### PR TITLE
[#6404] Server process removes leftover shared memory files on startup (main)

### PR DESCRIPTION
**Pull request by korydraughn**

### CODE REVIEW

File: server/main_server/src/rodsServer.cpp [hunk 1]
```diff
@@ -271,6 +271,68 @@ namespace
...
+                    }
+
+                    paths_to_remove.clear();
+
+                    // Build up a list of files to remove.
+                    // This avoids potential issues with the directory iterator.
+                    for (auto&& shm_file : fs::directory_iterator{shm_dir}) {
+                        const auto filename = shm_file.path().filename();
+                        const std::string_view s = filename.c_str();
+                        const std::string_view t = "irods";
```
Comment (trel) [inline server/main_server/src/rodsServer.cpp:306]: consider "irods_" instead... seems a little safer / more precise.

File: server/main_server/src/rodsServer.cpp [hunk 2]
```diff
@@ -271,6 +271,68 @@ namespace
...
+                        const auto e = std::end(s);
+
+                        if (std::search(std::begin(s), e, std::begin(t), std::end(t)) != e) {
```
Comment (trel) [inline server/main_server/src/rodsServer.cpp:309]: does this assert that it begins with "irods_"... or just that it is in there somewhere?

i think we want to assert 'begins with'.

File: server/main_server/src/rodsServer.cpp [hunk 3]
```diff
@@ -945,6 +1007,11 @@ int main(int argc, char** argv)
         return 1;
     }
 
+    // This must always be called after successfully creating the PID file, but before any
+    // libraries that handle shared memory. Violating these requirements can result in incorrect
+    // behavior depending on the whether an iRODS instance is already running.
```
Comment (trel) [inline server/main_server/src/rodsServer.cpp:1012]: "on the whether" -> "on whether"

File: server/main_server/src/rodsServer.cpp [hunk 4]
```diff
@@ -945,6 +1007,11 @@ int main(int argc, char** argv)
...
+    remove_leftover_shared_memory_files();
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply server/main_server/src/rodsServer.cpp:1013]: We have logic for cleaning up things on server stop, but they aren't guaranteed to always run.

[the remark this answers is not in the document]
Comment (trel) [reply server/main_server/src/rodsServer.cpp:1010]: right - cool.

**Comment by trel** [comment]
This will remove all the shm files that match - including ones that might be in use by another iRODS server running on the same machine... (non-package-install)...

So, we're going to hold on this for now to better determine why we are occasionally seeing collisions with these shm files - and solve that more directly, eventually.

**Comment by korydraughn** [comment]
This PR is no longer necessary now that #7988 has been merged.

Linking to #7229 so we can find this PR later.

Cleanup of shared memory files has been absorbed into the server directly. Plugins such as the S3 resource plugin can now use the setup/teardown functions for managing shared memory.

Closing.

**Comment by korydraughn** [comment]
I mispoke. The setup/teardown operations only apply to rule engine plugins. We'll need to provide similar functionality for resource plugins.

So, two issues are needed. One for the iRODS 5 server to grow support for setup/teardown for resource plugins. Another issue for the S3 resource plugin to implement those operations.

**Comment by korydraughn** [comment]
Hmm, @JustinKyleJames do all s3 resources share the same shared memory instance on a single iRODS server?

**Comment by korydraughn** [comment]
Thanks. I'll write up issues for this soon.

**Comment by alanking** [comment]
Issues have been filed...

Audit AMQP: https://github.com/irods/irods_rule_engine_plugin_audit_amqp/issues/160
Metadata guard: https://github.com/irods/irods_rule_engine_plugin_metadata_guard/issues/89
Logical quotas: https://github.com/irods/irods_rule_engine_plugin_logical_quotas/issues/121
PREP: https://github.com/irods/irods_rule_engine_plugin_python/issues/230
Indexing: https://github.com/irods/irods_capability_indexing/issues/166
Storage Tiering: https://github.com/irods/irods_capability_storage_tiering/issues/283
Resource plugins (investigation): https://github.com/irods/irods/issues/8072

**Pull request closed without merging**